audio_helper/src/openxtv4v.c

```diff
@@ -0,0 +1,382 @@
+///
+int openxt_v4v_send(V4VConnection *conn, V4VPacket *packet)
+{
+    // Local variables
+    int ret;
+
+    // Sanity checks
+    openxt_checkp(conn, -EINVAL);
+    openxt_checkp(packet, -EINVAL);
+    openxt_assert(openxt_v4v_get_length(packet) <= V4V_MAX_PACKET_BODY_SIZE, -EOVERFLOW);
+
+    // Make sure that we are actually connected. Note that we make this a
+    // quite failure because if a problem happens, we will close the connection
+    // and it's possible that the code might continue attempting to send, and
+    // we do not want to kill performance by logging a ton of error messages
+    openxt_assert_quiet(openxt_v4v_isconnected(conn) == true, -ENODEV);
+
+    // Send the packet. Note that we handle printing useful error messages
+    // here. All the user should have to do, is validate that the send was
+    // successful
+    ret = v4v_sendto(conn->fd, (char *)packet, packet->header.length, 0, &conn->remote_addr);
+    if (ret <= 0) {
+
+        switch (ret) {
+
+            // Failed to send anything
+            case 0:
+            {
+                openxt_warn("failed openxt_v4v_send, wrote 0 bytes: %d - %s\n", errno, strerror(errno));
+                openxt_v4v_close_internal(conn);
+                return -errno;
+            }
+
+            // Error
+            default:
+            {
+                openxt_warn("failed openxt_v4v_send: %d - %s\n", errno, strerror(errno));
+                openxt_v4v_close_internal(conn);
+                return -errno;
+            }
+        }
+    }
+
+    // Success
+    return ret - sizeof(V4VPacketHeader);
+}
+
+///
+/// The following function will send a V4V packet.
+///
+/// @param conn the V4V connection created using openxt_v4v_open.
+/// @param packet the packet to send
+///
+/// @return -EINVAL if conn or packet == NULL,
+///         -EIO if the received packet length != length in header,
+///         -ENODEV if conn is closed,
+///          negative errno if v4v_recvfrom fails,
+///          ret >= 0 on success representing number of bytes received
+///
+int openxt_v4v_recv(V4VConnection *conn, V4VPacket *packet)
+{
+    // Local variables
+    int ret;
+
+    // Sanity checks
+    openxt_checkp(conn, -EINVAL);
+    openxt_checkp(packet, -EINVAL);
+
+    // Make sure that we are actually connected. Note that we make this a
+    // quite failure because if a problem happens, we will close the connection
+    // and it's possible that the code might continue attempting to send, and
+    // we do not want to kill performance by logging a ton of error messages
+    openxt_assert_quiet(openxt_v4v_isconnected(conn) == true, -ENODEV);
+
+    // Send the packet. Note that we handle printing useful error messages
+    // here. All the user should have to do, is validate that the send was
+    // successful
+    ret = v4v_recvfrom(conn->fd, (char *)packet, sizeof(V4VPacket), 0, &conn->remote_addr);
+    if (ret <= 0) {
+
+        switch (ret) {
```

Comment:
It's there so that I can handle if there is an error, and then if there is an error, I want to different printout if the error was a linux error code, or 0 bytes written.